Add metadata selector and color by metadata on phylo tree.

MarkAZhang opened the pull request:
I decided to add a search box to `<Dropdown>` instead of using `<SearchBoxList>`, because `<SearchBoxList>` currently only selects multiple options (so I would have had to change code either way). Also, having a search box on our Dropdowns will be very useful. 

Because of this, the styling is a little different from `<SearchBoxList>`.

-----

<img width="944" alt="screen shot 2018-12-07 at 3 58 13 pm" src="https://user-images.githubusercontent.com/837004/49678520-a3069e80-fa39-11e8-87ad-58b0c9b8a55a.png">

<img width="535" alt="screen shot 2018-12-07 at 3 58 30 pm" src="https://user-images.githubusercontent.com/837004/49678521-a69a2580-fa39-11e8-9bdf-28a37e1bc1d3.png">

<img width="439" alt="screen shot 2018-12-07 at 3 58 40 pm" src="https://user-images.githubusercontent.com/837004/49678522-a7cb5280-fa39-11e8-9b7b-bd045442c79d.png">

<img width="942" alt="screen shot 2018-12-07 at 3 58 46 pm" src="https://user-images.githubusercontent.com/837004/49678528-aef26080-fa39-11e8-9b49-ae2ffec478f2.png">

jshoe commented:
- Left margin missing? or is that intentional

![screen shot 2018-12-10 at 10 56 29 am](https://user-images.githubusercontent.com/5652739/49754682-9a95aa00-fc6b-11e8-988e-4340ebdf0c55.png)

- Minor thing that's hard to notice: It looks like "RELATIVE DISTANCE" gets redrawn (and builds up) when you switch around the Color By
![screen shot 2018-12-10 at 10 57 45 am](https://user-images.githubusercontent.com/5652739/49754712-aaad8980-fc6b-11e8-86fd-e66b79b940a9.png)

- Would be nice to get "arrow keys" on the search box dropdown someday

- Basically ignoring the code-move changes in the diff

MarkAZhang commented:
<img width="693" alt="screen shot 2018-12-20 at 6 56 09 pm" src="https://user-images.githubusercontent.com/837004/50322182-ed9c0800-0488-11e9-8e19-f6b3f54f1814.png">

Update takes new dropdown styling into account.

Also added a search box and filtering functionality to the base BareDropdown component, so we can easily add a search box to any of our dropdowns.

Also fixed a bug in `metadatum.rb` that was causing all metadata values to be returned as `""`. The issue was that `string_validated_value`, `date-validated_value`, etc. were not instance variables of the model. Verified that the new solution worked.

jshoe commented on app/models/metadatum.rb:
```diff
@@ -352,7 +352,7 @@ def self.new_without_save(sample, key, value)
 
   def validated_value
-    return instance_variable_get("#{data_type}_validated_value")
+    return self["#{data_type}_validated_value"]
```
wait, was this change really necessary? isn't it basically doing the same thing (getting the value of the instance variable?)

MarkAZhang replied:
(Posting this mainly for posterity)

I did some research. It's surprisingly confusing.

`instance_variable_get("@string_validated_value")` (you need the `@`, which was missing in the code) will get the instance variable `@string_validated_value`. Unfortunately, the database value for `string_validated_value` is not stored on `@string_validated_value`, but actually on `@attributes["string_validated_value"]` (Rails magic). And if you try to access `@attributes["string_validated_value"]` directly, you'll find that it's not actually the database value,  but an internal object of type `ActiveRecord::Attribute::FromDatabase`. I think Rails intends you to call the getter method `string_validated_value` it automatically generates, via self.string_validated_value, which gives you the correct value. 

NOTE:
Instance variables are always private by default. Within a class, instance variables can only be accessed by `@foo`. If you instead type `foo`, it will first try to resolve to a local variable, and then resolve to `self.foo`, which is the getter function for `foo`. This is why  `data_type` works within the class. It is actually resolving to `self.data_type` which is the Rails auto-generated getter method for the database attribute. `@data_type` would NOT work, because that's the instance variable, and the database value is not stored there.
